src/Catalyst.Core.Lib/Network/Ip.cs

```diff
@@ -0,0 +1,150 @@
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Catalyst.Node. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Sockets;
+using System.Reactive.Linq;
+using System.Reflection;
+using System.Threading.Tasks;
+using Catalyst.Core.Lib.Util;
+using Serilog;
+
+namespace Catalyst.Core.Lib.Network
+{
+    public static class Ip
+    {
+        private static readonly ILogger Logger = Log.Logger.ForContext(MethodBase.GetCurrentMethod().DeclaringType);
+
+        public static readonly ReadOnlyCollection<string> DefaultIpEchoUrls;
+
+        static Ip()
+        {
+            DefaultIpEchoUrls = new List<string>
+            {
+                "https://api.ipify.org",
+                "https://ipecho.net/plain",
+                "https://ifconfig.co/ip",
+                "https://ipv4bot.whatismyipaddress.com"
+            }.AsReadOnly();
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <returns></returns>
+        public static async Task<IPAddress> GetPublicIpAsync(IObservable<string> ipEchoUrls = null)
+        {
+            var defaultedIpEchoUrls = ipEchoUrls ?? DefaultIpEchoUrls.ToObservable();
+
+            var echoedIp = await defaultedIpEchoUrls
+               .Select((url, i) => Observable.FromAsync(async () => await TryGetExternalIpFromEchoUrlAsync(url).ConfigureAwait(false)))
+               .Merge()
+               .FirstAsync(t => t != null);
+
+            return echoedIp;
+        }
+
+        /// <summary>
+        ///     Creates a standardised format byte array that can handle a IPv6 address or an IPv4 with leading bytes padded with 0x0
+        /// </summary>
+        /// <param name="address"></param>
+        /// <returns></returns>
+        public static byte[] To16Bytes(this IPAddress address)
+        {
+            var ipChunk = ByteUtil.InitialiseEmptyByteArray(16);
+            var ipBytes = address.GetAddressBytes();
+
+            if (ipBytes.Length == 4)
+            {
+                Buffer.BlockCopy(ipBytes, 0, ipChunk, 12, 4);
+            }
+            else
+            {
+                ipChunk = ipBytes;
+            }
+
+            Logger.Verbose(string.Join(" ", ipChunk));
+
+            return ipChunk;
+        }
+
+        private static async Task<IPAddress> TryGetExternalIpFromEchoUrlAsync(string url)
+        {
+            try
+            {
+                var req = WebRequest.Create(url);
+                using (var response = await req.GetResponseAsync())
+                using (var reader =
+                    new StreamReader(response.GetResponseStream() ?? throw new InvalidOperationException()))
+                {
+                    var responseContent = (await reader.ReadToEndAsync()).Trim();
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.](https://app.codacy.com/app/catalyst-network/Catalyst.Node/pullRequest?prid=4318201)